File: codes/matlab/ch16/trap_band_limited.m
% trap_band_limited.m
% Chapter 16: Integration of Periodic Functions
%
% Computational Etude 16.2: Band-Limited Exactness and One-Mode Aliasing.
%
% (a) The N-point periodic trapezoidal rule is exact for any
%     trigonometric polynomial of degree m < N.
% (b) For a single mode cos(k*theta) with N fixed, the rule still
%     gives the correct value of zero unless k is an integer multiple
%     of N -- even when k > N/2 (the aliased Nyquist regime).
%
% Generates Figure 16.2: band_limited.pdf  (two-panel figure)
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Etudes: A Spectral Approach"
% https://github.com/dutykh/computational-etudes

clear; close all; clc;

set(0, 'DefaultAxesFontSize', 11);
set(0, 'DefaultTextFontSize', 11);
set(0, 'DefaultLineLineWidth', 1.5);
set(0, 'DefaultAxesLineWidth', 0.8);

NAVY  = [20, 45, 110] / 255;
CORAL = [231, 76, 60] / 255;

output_dir = fullfile(fileparts(mfilename('fullpath')), ...
    '..', '..', '..', 'textbook', 'figures', 'ch16', 'matlab');
if ~exist(output_dir, 'dir'), mkdir(output_dir); end

% -------------------------------------------------------------------------
% Part (a): random trigonometric polynomial of degree m = 10
% -------------------------------------------------------------------------
rng(42);
m = 10;
a = randn(m + 1, 1);
b = randn(m + 1, 1);

% trig_poly(theta) = a(1) + sum_{k=1}^m [a(k+1)*cos(k*theta) + b(k+1)*sin(k*theta)]
trig_poly = @(t) a(1) + sum(a(2:end) .* cos((1:m)' * t(:)') + ...
                            b(2:end) .* sin((1:m)' * t(:)'), 1);

I_exact = 2 * pi * a(1);
N_values_a = 4:30;
errors_a = zeros(size(N_values_a));
for i = 1:length(N_values_a)
    N = N_values_a(i);
    theta = 2*pi*(0:N-1)/N;
    errors_a(i) = abs((2*pi/N) * sum(trig_poly(theta)) - I_exact);
end
errors_a = max(errors_a, 1e-17);

fprintf('Part (a): random trig polynomial of degree m = %d\n', m);
fprintf('Exact integral: 2*pi*a(1) = %.10f\n', I_exact);
fprintf('Error at N = m = %d: %.4e\n', m, errors_a(m - 3));
fprintf('Error at N = m+1 = %d: %.4e\n\n', m + 1, errors_a(m - 2));

% -------------------------------------------------------------------------
% Part (b): single mode cos(k*theta), N = 16 fixed, sweep k
% -------------------------------------------------------------------------
N_fixed = 16;
k_values = 0:32;
errors_b = zeros(size(k_values));
for i = 1:length(k_values)
    k = k_values(i);
    theta = 2*pi*(0:N_fixed-1)/N_fixed;
    if k == 0
        I_exact_k = 2*pi;
    else
        I_exact_k = 0;
    end
    errors_b(i) = abs((2*pi/N_fixed) * sum(cos(k * theta)) - I_exact_k);
end
errors_b = max(errors_b, 1e-17);

% -------------------------------------------------------------------------
% Two-panel plot
% -------------------------------------------------------------------------
fig = figure('Position', [100, 100, 1100, 450]);

subplot(1, 2, 1);
semilogy(N_values_a, errors_a, 'o-', 'Color', CORAL, 'MarkerSize', 5, ...
    'LineWidth', 1.1, 'MarkerFaceColor', CORAL);
hold on;
xline(m, '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.9);
hold off;
xlabel('Number of nodes $N$', 'Interpreter', 'latex');
ylabel('Absolute error $|I_N - I|$', 'Interpreter', 'latex');
title('(a) trigonometric polynomial of degree $m = 10$', 'Interpreter', 'latex');
grid on; set(gca, 'GridAlpha', 0.3);
xlim([3, 31]); ylim([1e-18, 1e2]);

subplot(1, 2, 2);
% Bar plot with separate colours for the failure modes
mask_fail = errors_b > 1e-10;
bar(k_values(~mask_fail), errors_b(~mask_fail), 'FaceColor', NAVY, 'EdgeColor', 'none');
hold on;
bar(k_values(mask_fail), errors_b(mask_fail), 'FaceColor', CORAL, 'EdgeColor', 'none');
xline(N_fixed, '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.9);
xline(2*N_fixed, '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.9);
hold off;
xlabel('Frequency $k$', 'Interpreter', 'latex');
ylabel('$|I_N(\cos k\theta) - I(\cos k\theta)|$', 'Interpreter', 'latex');
title('(b) single modes, fixed $N = 16$', 'Interpreter', 'latex');
set(gca, 'YScale', 'log');
ylim([1e-17, 1e1]); xlim([-0.5, 32.5]);
grid on; set(gca, 'GridAlpha', 0.3);

exportgraphics(gcf, fullfile(output_dir, 'band_limited.pdf'), 'ContentType', 'vector');
exportgraphics(gcf, fullfile(output_dir, 'band_limited.png'), 'Resolution', 300);
fprintf('Figure saved to %s\n', fullfile(output_dir, 'band_limited.pdf'));

File: codes/julia/ch16/trap_band_limited.jl
#!/usr/bin/env julia
#=
trap_band_limited.jl
Chapter 16: Integration of Periodic Functions

Computational Etude 16.2: Band-Limited Exactness and One-Mode Aliasing.

(a) The N-point periodic trapezoidal rule is exact for any
    trigonometric polynomial of degree m < N.
(b) For a single mode cos(k*theta) with N fixed, the rule still
    gives the correct value of zero unless k is an integer multiple
    of N -- even when k > N/2 (the aliased Nyquist regime).

Generates Figure 16.2: band_limited.pdf  (two-panel figure)

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
=#

using Plots
using LaTeXStrings
using Random

gr()

const NAVY  = RGB(20/255, 45/255, 110/255)
const CORAL = RGB(231/255, 76/255, 60/255)

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch16", "julia")
mkpath(OUTPUT_DIR)

trapezoidal_periodic(f, N) = (2π/N) * sum(f.(2π * (0:N-1) / N))

function main()
    # ----- Part (a): random trigonometric polynomial of degree m = 10 -----
    Random.seed!(42)
    m = 10
    a = randn(m + 1)
    b = randn(m + 1)
    function trig_poly(θ)
        s = a[1]
        for k in 1:m
            s += a[k+1] * cos(k * θ) + b[k+1] * sin(k * θ)
        end
        return s
    end
    I_exact = 2π * a[1]

    N_values_a = 4:30
    errors_a = [abs(trapezoidal_periodic(trig_poly, N) - I_exact) for N in N_values_a]
    errors_a = max.(errors_a, 1e-17)

    println("Part (a): random trig polynomial of degree m = $m")
    println("Exact integral 2π*a[1] = ", I_exact)
    println("Error at N = $(m): ", errors_a[m - 3])
    println("Error at N = $(m + 1): ", errors_a[m - 2])

    # ----- Part (b): single mode cos(k*theta), N = 16 fixed, sweep k -----
    N_fixed = 16
    k_values = 0:32
    errors_b = zeros(length(k_values))
    for (i, k) in enumerate(k_values)
        θ = 2π * (0:N_fixed-1) / N_fixed
        I_exact_k = (k == 0) ? 2π : 0.0
        errors_b[i] = abs((2π/N_fixed) * sum(cos.(k .* θ)) - I_exact_k)
    end
    errors_b = max.(errors_b, 1e-17)

    # ----- Two-panel plot -----
    p1 = plot(collect(N_values_a), errors_a,
              yscale = :log10,
              xlabel = L"Number of nodes $N$",
              ylabel = L"Absolute error $|I_N - I|$",
              title = L"(a) trig polynomial of degree $m = 10$",
              label = "", marker = :circle, markersize = 5, color = CORAL,
              linewidth = 1.1, grid = true, gridalpha = 0.3,
              framestyle = :box, xlims = (3, 31), ylims = (1e-18, 1e2))
    vline!(p1, [m], linestyle = :dash, color = RGB(0.5, 0.5, 0.5),
           linewidth = 0.9, label = "")

    bar_colors = [errors_b[i] > 1e-10 ? CORAL : NAVY for i in 1:length(k_values)]
    p2 = bar(collect(k_values), errors_b,
             yscale = :log10,
             xlabel = L"Frequency $k$",
             ylabel = L"$|I_N(\cos k\theta) - I(\cos k\theta)|$",
             title = L"(b) single modes, fixed $N = 16$",
             label = "", color = bar_colors, linecolor = bar_colors,
             grid = true, gridalpha = 0.3, framestyle = :box,
             xlims = (-0.5, 32.5), ylims = (1e-17, 1e1))
    vline!(p2, [N_fixed, 2*N_fixed], linestyle = :dash,
           color = RGB(0.5, 0.5, 0.5), linewidth = 0.9, label = "")

    p = plot(p1, p2, layout = (1, 2), size = (1100, 450))
    savefig(p, joinpath(OUTPUT_DIR, "band_limited.pdf"))
    savefig(p, joinpath(OUTPUT_DIR, "band_limited.png"))
    println("\nFigure saved to ", joinpath(OUTPUT_DIR, "band_limited.pdf"))
end

main()
